atomix/utils/src/main/java/io/atomix/utils/serializer/FallbackNamespace.java

```diff
@@ -0,0 +1,128 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.atomix.utils.serializer;
+
+import static org.slf4j.LoggerFactory.getLogger;
+
+import com.google.common.collect.ImmutableList;
+import io.atomix.utils.serializer.NamespaceImpl.Builder;
+import io.atomix.utils.serializer.NamespaceImpl.RegistrationBlock;
+import java.nio.ByteBuffer;
+import org.slf4j.Logger;
+
+public class FallbackNamespace implements Namespace {
+
+  private static final Logger LOG = getLogger(FallbackNamespace.class);
+  private static final String DESERIALIZE_ERROR =
+      "Deserialization failed with both the versioned and fallback serializers. The fallback serializer failed with:\n %s";
+  private final Namespace fallback;
+  private final Namespace namespace;
+
+  FallbackNamespace(final Namespace fallback, final Namespace namespace) {
+    this.fallback = fallback;
+    this.namespace = namespace;
+  }
+
+  public FallbackNamespace(final NamespaceImpl.Builder builder) {
+    final Builder copy = builder.copy();
+    this.fallback = builder.build();
+    this.namespace = copy.name(copy.getName() + "-compatible").setCompatible(true).build();
+  }
+
+  /**
+   * Serializes given object to byte array using Kryo instance in pool.
+   *
+   * <p>Note: Serialized bytes must be smaller than {@link NamespaceImpl#MAX_BUFFER_SIZE}.
+   *
+   * @param obj Object to serialize
+   * @return serialized bytes
+   */
+  public byte[] serialize(final Object obj) {
+    return fallback.serialize(obj);
+  }
+
+  /**
+   * Serializes given object to byte array using Kryo instance in pool.
+   *
+   * @param obj Object to serialize
+   * @param bufferSize maximum size of serialized bytes
+   * @return serialized bytes
+   */
+  public byte[] serialize(final Object obj, final int bufferSize) {
+    return fallback.serialize(obj, bufferSize);
+  }
+
+  /**
+   * Serializes given object to byte buffer using Kryo instance in pool.
+   *
+   * @param obj Object to serialize
+   * @param buffer to write to
+   */
+  public void serialize(final Object obj, final ByteBuffer buffer) {
+    fallback.serialize(obj, buffer);
+  }
+
+  /**
+   * Deserializes given byte array to Object using Kryo instance in pool.
+   *
+   * @param bytes serialized bytes
+   * @param <T> deserialized Object type
+   * @return deserialized Object
+   */
+  public <T> T deserialize(final byte[] bytes) {
+    try {
+      return namespace.deserialize(bytes);
```

Comment:
A broker running this version always serialize with `fallback`, but deserialize with the new one. Is that ideal? A cluster running with all 0.24.x broker then always have to fallback when deserializing. Not sure how much performance impact it will have.

Reply:
I think that namespace should be the current one and fallback should be the new one because we have to use fallback only during rolling upgrade to 0.25.x.

Reply:
That was the original idea but some objects serialized with the CompatibleSerializer didn't fail when being deserialized with the FieldSerializer, instead being deserialized to a bad object which caused NPEs. So I had to switch the order. I realize it's not the ideal path but at least I didn't see a meaningful impact in the benchmark I ran

Reply:
Ok :+1: